Show compilation error in submission-result if submitting to server returns compile failed

<h3>Compile failed object returned:</h3>
api_version:7
all_tests_passed:false
user_id:2608
login:"testi"
course:"hy-tira-k20"
exercise_name:"viikko04-Viikko04Tehtava4"
status:"error"
points:Array[0]
validations:Object
valgrind:""
submission_url:"https://tmc.mooc.fi/submissions/5908450"
solution_url:null
submitted_at:"2020-02-06T18:53:46.273+02:00"
processing_time:9
reviewed:false
requests_review:false
paste_url:null
message_for_paste:null
missing_review_points:Array[0]
error:"Compilation error:
stdout: 
-pre-init:

-init-private:

-init-user:

-init-project:

-init-macrodef-property:

-do-init:

-post-init:

-init-check:

-init-ap-cmdline-properties:

-init-macrodef-javac-with-processors:

-init-macrodef-javac-without-processors:

-init-macrodef-javac:

-init-macrodef-test-impl:

-init-macrodef-junit-init:

-init-macrodef-junit-single:

-init-test-properties:

-init-macrodef-junit-batch:

-init-macrodef-junit:

-init-macrodef-junit-impl:
Trying to override old definition of task http://www.netbeans.org/ns/j2se-project/3:test-impl

-init-macrodef-testng:

-init-macrodef-testng-impl:

-init-macrodef-test:

-init-macrodef-junit-debug:

-init-macrodef-junit-debug-batch:

-init-macrodef-junit-debug-impl:

-init-macrodef-test-debug-junit:

-init-macrodef-testng-debug:

-init-macrodef-testng-debug-impl:

-init-macrodef-test-debug-testng:

-init-macrodef-test-debug:

-init-debug-args:

-init-macrodef-nbjpda:

-init-macrodef-debug:

-init-macrodef-java:

-init-presetdef-jar:

-init-ap-cmdline-supported:

-init-ap-cmdline:

init:

-deps-jar-init:

deps-jar:
    [mkdir] Created dir: /app/build

-warn-already-built-jar:
[propertyfile] Updating property file: /app/build/built-jar.properties

-check-automatic-build:

-clean-after-automatic-build:

-verify-automatic-build:

-pre-pre-compile:
    [mkdir] Created dir: /app/build/classes

-pre-compile:

-copy-persistence-xml:

-compile-depend:

-do-compile:
    [mkdir] Created dir: /app/build/empty
    [mkdir] Created dir: /app/build/generated-sources/ap-source-output
    [javac] Compiling 2 source files to /app/build/classes
    [javac] /app/src/Pinolaskin.java:5: error: missing return statement
    [javac]     }    
    [javac]     ^
    [javac] 1 error

stderr: 
BUILD FAILED
/app/nbproject/build-impl.xml:930: The following error occurred while executing this line:
/app/nbproject/build-impl.xml:270: Compile failed; see the compiler error output for details.

Total time: 2 seconds
"






<h3>Correct return if no compilation error:</h3>
api_version:7
all_tests_passed:false
user_id:2608
login:"testi"
course:"mooc-ohjelmoinnin-perusteet"
exercise_name:"osa01-Osa01_05.Viesti"
status:"fail"
points:Array[0]
validations:Object
valgrind:""
submission_url:"https://tmc.mooc.fi/submissions/5908817"
solution_url:null
submitted_at:"2020-02-06T19:12:38.981+02:00"
processing_time:11
reviewed:false
requests_review:false
paste_url:null
message_for_paste:null
missing_review_points:Array[0]
test_cases:Array[2]
cssPath:Object
bootstrapPath:Object

--- a/resources/templates/submission-result.html
+++ b/resources/templates/submission-result.html
@@ -8,18 +8,16 @@
 <body>
     <div class="container">
         <div class="row py-2">
-            {{#if all_tests_passed}}
-            <h1 class="passed-header">All tests passed on the server</h1>
-            {{else}}
-            <h1>Some tests failed on the server</h1>
-            {{/if}}
+            {{{submission_status this}}}
         </div>
-        <div class="row my-1">
-            <h3>Points gained: {{points}}</h3>
-        </div>
-        <div class="row my-1">
-            {{{progress_bar test_cases}}}
-        </div>
+        {{#if test_cases}}
+            <div class="row my-1">
+                <h3>Points gained: {{points}}</h3>
+            </div>
+            <div class="row my-1">
+                    {{{progress_bar test_cases}}}
+            </div>
+        {{/if}}
         {{#if feedback_questions}}
         <div id="feedback" class="my-5">
             <h4>Anna palautetta</h4>
@@ -68,7 +66,7 @@
                         </div>
                         <div class="row">
                             <div class="col">
-                                <span class="message">{{this.message}}</span>
+                                <pre>{{this.message}}</pre>
                             </div>
                         </div>
                     </div>

--- a/src/api/types.ts
+++ b/src/api/types.ts
@@ -116,7 +116,7 @@ export type SubmissionResultReport = {
     login: string;
     course: string;
     exercise_name: string;
-    status: "fail" | "ok";
+    status: "fail" | "ok" | "error";
     points: string[];
     validations: any;
     valgrind: string;
@@ -129,7 +129,7 @@ export type SubmissionResultReport = {
     paste_url: string | null;
     message_for_paste: string | null;
     missing_review_points: string[];
-    test_cases: Array<{
+    test_cases?: Array<{
         name: string;
         successful: boolean;
         message: string;
@@ -138,6 +138,7 @@ export type SubmissionResultReport = {
     }>;
     feedback_questions?: SubmissionFeedbackQuestion[];
     feedback_answer_url?: string;
+    error?: string;
 };
 
 export type SubmissionFeedbackQuestion = {

--- a/src/ui/templateEngine.ts
+++ b/src/ui/templateEngine.ts
@@ -51,9 +51,9 @@ export default class TemplateEngine {
             if (results.status === "ok") {
                 if (results.all_tests_passed) {
                     return `<h1 class="passed-header">All tests passed on the server</h1>`;
-                } else {
-                    return `<h1>Some tests failed on the server</h1>`;
                 }
+            } else if (results.status === "fail") {
+                return `<h1>Some tests failed on the server</h1>`;
             } else if (results.status === "error") {
                 return `<h1>Server returned following error:
                         <pre>${results.error}</pre>`;

--- a/src/ui/templateEngine.ts
+++ b/src/ui/templateEngine.ts
@@ -51,9 +51,9 @@ export default class TemplateEngine {
             if (results.status === "ok") {
                 if (results.all_tests_passed) {
                     return `<h1 class="passed-header">All tests passed on the server</h1>`;
+                } else {
+                    return `<h1>Some tests failed on the server</h1>`;
                 }
-            } else if (results.status === "fail") {
-                return `<h1>Some tests failed on the server</h1>`;
             } else if (results.status === "error") {
                 return `<h1>Server returned following error:
                         <pre>${results.error}</pre>`;

--- a/src/ui/templateEngine.ts
+++ b/src/ui/templateEngine.ts
@@ -3,7 +3,7 @@ import * as handlebars from "handlebars";
 import * as path from "path";
 import * as vscode from "vscode";
 
-import { TmcLangsTestResult } from "../api/types";
+import { TmcLangsTestResult, SubmissionResultReport } from "../api/types";
 import Resources from "../config/resources";
 
 export default class TemplateEngine {
@@ -44,6 +44,19 @@ export default class TemplateEngine {
             } else {
                 // TODO: Parse COMPILE_FAILED error logs
                 return "<h1>Something went wrong while running the tests</h1>";
+            }
+        });
+
+        handlebars.registerHelper("submission_status", (results: SubmissionResultReport) => {
+            if (results.status === "ok") {
+                if (results.all_tests_passed) {
+                    return `<h1 class="passed-header">All tests passed on the server</h1>`;
+                } else {
+                    return `<h1>Some tests failed on the server</h1>`;
+                }
+            } else if (results.status === "error") {
+                return `<h1>Server returned following error:
+                        <pre>${results.error}</pre>`;
             }
         });
         /**

--- a/src/ui/templateEngine.ts
+++ b/src/ui/templateEngine.ts
@@ -3,7 +3,7 @@ import * as handlebars from "handlebars";
 import * as path from "path";
 import * as vscode from "vscode";
 
-import { TmcLangsTestResult, SubmissionResultReport } from "../api/types";
+import { SubmissionResultReport, TmcLangsTestResult } from "../api/types";
 import Resources from "../config/resources";
 
 export default class TemplateEngine {

--- a/src/ui/templateEngine.ts
+++ b/src/ui/templateEngine.ts
@@ -51,18 +51,22 @@ export default class TemplateEngine {
             }
         });
 
+        /**
+         * Submission result show correct heading or compilation error
+         */
         handlebars.registerHelper("submission_status", (results: SubmissionResultReport) => {
             if (results.status === "ok") {
                 if (results.all_tests_passed) {
                     return `<h1 class="passed-header">All tests passed on the server</h1>`;
-                } else {
-                    return `<h1>Some tests failed on the server</h1>`;
                 }
+            } else if (results.status === "fail") {
+                    return `<h1>Some tests failed on the server</h1>`;
             } else if (results.status === "error") {
                 return `<h1>Server returned following error:
-                        <pre>${results.error}</pre>`;
+                        <pre style="font-size: 14px">${results.error}</pre>`;
             }
         });
+
         /**
          * Progress bar for running tests and submission.
          */

